### PR TITLE
Add girder_worker job cancellation

### PR DESCRIPTION
**Pull request by cjh1**
The brother PR in girder_worker is girder/girder_worker#128

### CODE REVIEW

File: plugins/worker/server/__init__.py [hunk 1]
```diff
@@ -91,6 +94,32 @@ def schedule(event):
...
+
+        if celeryTaskId is None:
+            msg = ("Unable to cancel Celery task. Job '%s' doesn't have a Celery task id."
+                   % job['_id'])
+            logger.warn(msg)
+            return
+
+        # Send the task to celery
+        asyncResult = AsyncResult(celeryTaskId)
+        asyncResult.revoke()
```
Comment (cjh1) [inline plugins/worker/server/__init__.py]: @kotfic I'm not appose to added a terminate and signal parameter to this endpoint, which I think is all that would required for your use case? With the appropriate warnings of course :smile:

**Comment by cjh1** [comment]
Still todo:

- [ ] Update canceling icon to rotating orange
- [ ] Make canceling jobs in the following states a noop: ['success', 'error', 'canceling']

**Comment by zachmullen** [comment]
The rotation of the icon is super easy: https://github.com/girder/girder/blob/master/plugins/jobs/web_client/JobStatus.js#L75

**Comment by cjh1** [comment]
Thanks, super easy sounds good :smile:

File: plugins/jobs/web_client/views/JobDetailsWidget.js [hunk 2]
```diff
@@ -112,6 +120,19 @@ var JobDetailsWidget = View.extend({
...
+
+    _cancelJob: function () {
+        const jobId = this.job.id;
+        restRequest({
+            path: `job/${jobId}/cancel`,
+            type: 'PUT',
+            error: null
+        }).done(_.bind(function () {
+            console.log('success');
+        }, this)).error(_.bind(function (err) {
```
Comment (brianhelba) [inline plugins/jobs/web_client/views/JobDetailsWidget.js]: Per the changes in https://github.com/girder/girder/pull/2032 , this should be changed to `.fail`.

**Comment by zachmullen** [comment]
@cjh1 The failing test appears to be a result of a bad request, see this screenshot: http://my.cdash.org/testDetails.php?test=32164915&build=1223795

**Comment by zachmullen** [comment]
[The JSON](https://www.celebsclothing.com/product_images/e/183/Happy_Days_Jacket__82354_std.jpg)

**Comment by cjh1** [comment]
@zachmullen Thanks that helps alot, I will fix up those tests.

**Comment by cjh1** [comment]
@girder/developers This is ready for review, PTAL

File: plugins/jobs/web_client/views/JobListWidget.js [hunk 3]
```diff
@@ -260,7 +307,33 @@ var JobListWidget = View.extend({
             filter.statuses = JSON.stringify(this.statusFilter);
         }
         this.collection.params = filter;
+
         return this.collection.fetch({}, true);
+    },
+
+    _cancelJobs: function () {
+        Promise.all(
```
Comment (brianhelba) [inline plugins/jobs/web_client/views/JobListWidget.js]: Per the consensus in #2068, we need to stick with jQuery Promise until at least Girder 3.

File: plugins/jobs/web_client/views/JobListWidget.js [hunk 4]
```diff
@@ -260,7 +307,33 @@ var JobListWidget = View.extend({
...
+            Object.keys(this.jobCheckedStates)
+                .filter(jobId => {
+                    var status = this.jobs.find(job => job.id === jobId).get('status');
+                    return [JobStatus.CANCELED, JobStatus.SUCCESS, JobStatus.ERROR].indexOf(status) === -1;
+                })
+                .map(jobId => Promise.resolve(restRequest({
```
Comment (brianhelba) [inline plugins/jobs/web_client/views/JobListWidget.js]: Per the consensus in #2068, we need to stick with jQuery Promise until at least Girder 3.

File: plugins/jobs/plugin_tests/jobsSpec.js [hunk 5]
```diff
@@ -59,12 +59,28 @@ $(function () {
...
 
                 expect($('.g-timeline-segment').length).toBe(3);
                 expect($('.g-timeline-point').length).toBe(4);
                 expect($('.g-timeline-start-label').text()).toBe('0 s');
                 expect($('.g-timeline-end-label').text()).toBe('12 s');
-                expect($('.g-timeline-point')[3].className).toContain('g-job-color-success');
+
+                function toHex(c) {
+                    var hex = parseInt(c, 10).toString(16);
+                    return hex.length == 1 ? '0' + hex : hex;
```
Comment (brianhelba) [inline plugins/jobs/plugin_tests/jobsSpec.js]: Always use `===` in JS.

File: plugins/jobs/web_client/views/JobListWidget.js [hunk 6]
```diff
@@ -45,6 +69,17 @@ var JobListWidget = View.extend({
             obj[status.text] = true;
             return obj;
         }, {});
+        this.jobCheckedStates = {};
+        this.jobHighlightStates = {};
+        Object.defineProperty(this, 'jobs', {
```
Comment (brianhelba) [inline plugins/jobs/web_client/views/JobListWidget.js]: Backbone Views don't expose their methods this way. Why is this being used?

File: plugins/jobs/web_client/views/JobListWidget.js (hunk 3 above)
Comment (matthewma7) [inline plugins/jobs/web_client/views/JobListWidget.js]: I have another PR that has the Promise.all like $.whenAll, I am hoping that PR could go in before this one, then I'll merge from master to change this to jQuery Promise.

File: plugins/jobs/web_client/views/JobListWidget.js (hunk 6 above)
Comment (matthewma7) [inline plugins/jobs/web_client/views/JobListWidget.js]: This is not directly related to Backbone, this is the es5 way to define property without ES6 class syntax.

File: plugins/worker/server/__init__.py [hunk 7]
```diff
@@ -94,6 +117,34 @@ def schedule(event):
...
+            logger.warn(msg)
+            return
+
+        if job['status'] not in [JobStatus.SUCCESS, JobStatus.ERROR]:
+            # Set the job status to canceling
+            ModelImporter.model('job', 'jobs').updateJob(job, status=CustomJobStatus.CANCELING)
+
+            # Send the revoke request.
+            asyncResult = AsyncResult(celeryTaskId)
+            asyncResult.revoke()
```
Comment (kotfic) [inline plugins/worker/server/__init__.py]: ```signal``` and ```terminate``` kwargs are not being passed through from REST endpoint.

File: plugins/worker/server/__init__.py [hunk 8]
```diff
@@ -94,6 +117,34 @@ def schedule(event):
...
 
 
+def cancel(event):
+    """
+    This is bound to the "jobs.cancel" event, and will be triggered any time
+    a job is canceled. This handler will process any job that has the
+    handler field set to "worker_handler".
+    """
+    job = event.info
+    if job['handler'] == 'worker_handler':
```
Comment (kotfic) [inline plugins/worker/server/__init__.py]: ```worker_handler``` is the handler that manages traditional ```app.send_message(...)``` task production.  Please also white list ```"celery_handler"``` which is the handler type we've assigned to traditional celery based task distribution (e.g. with ```my_task.delay(...)``` and ```my_task.apply_async(...)```)

File: plugins/worker/server/__init__.py (hunk 7 above)
Comment (cjh1) [inline plugins/worker/server/__init__.py]: Correctly, I planned on adding that in a later PR if that makes sense?

File: plugins/worker/server/__init__.py (hunk 8 above)
Comment (cjh1) [inline plugins/worker/server/__init__.py]: Ok, will do

**Comment by cjh1** [comment]
This is ready for another look, tests are passing. Not sure how to read the coverage report, is this a fault positive?

File: plugins/jobs/web_client/views/JobListWidget.js (hunk 6 above)
Comment (brianhelba) [inline plugins/jobs/web_client/views/JobListWidget.js]: Unless there's an important reason for these attributes to be assigned this way, I'd strongly prefer that these be created using the normal Backbone syntax of `extend`. Also, these should be private (prefixed with a `_`).

File: plugins/jobs/web_client/views/JobDetailsWidget.js (hunk 2 above)
Comment (brianhelba) [inline plugins/jobs/web_client/views/JobDetailsWidget.js]: This still needs to be fixed.

File: plugins/jobs/web_client/views/JobListWidget.js [hunk 9]
```diff
@@ -232,20 +277,25 @@ var JobListWidget = View.extend({
...
         this._renderData();
-        this._highlightRecordIfOnList(event.data._id);
+        this.trySetHighlightRecord(event.data._id);
     },
 
     _jobCreated: function (event) {
         this._fetchWithFilter()
-            .done(() => {
-                this._highlightRecordIfOnList(event.data._id);
+            .then(() => {
```
Comment (brianhelba) [inline plugins/jobs/web_client/views/JobListWidget.js]: This should use `.done`, with no return value.

File: plugins/jobs/web_client/views/JobListWidget.js [hunk 10]
```diff
@@ -261,7 +311,34 @@ var JobListWidget = View.extend({
...
+                .filter((jobId) => {
+                    var jobModel = this.jobs.find((job) => job.id === jobId);
+                    return JobStatus.isCancelable(jobModel);
+                })
+                .map((jobId) => restRequest({
+                    path: `job/${jobId}/cancel`,
+                    type: 'PUT',
+                    error: null
+                }))
+        ).then((results) => {
```
Comment (brianhelba) [inline plugins/jobs/web_client/views/JobListWidget.js]: This should use `.done`, with no return value.

File: plugins/jobs/web_client/views/JobListWidget.js (hunk 6 above)
Comment (matthewma7) [inline plugins/jobs/web_client/views/JobListWidget.js]: @brianhelba Could you detail what do you mean by using Backbone `extend`?

Comment (matthewma7) [inline plugins/jobs/web_client/views/JobListWidget.js]: @brianhelba Could you illustrate your point on `extend`?

File: plugins/jobs/web_client/views/JobListWidget.js (hunk 9 above)
Comment (matthewma7) [inline plugins/jobs/web_client/views/JobListWidget.js]: I have to say the `done` and `then` mix use is really confusing to me. 😵

File: plugins/jobs/web_client/views/JobListWidget.js [hunk 11]
```diff
@@ -233,20 +269,24 @@ var JobListWidget = View.extend({
...
+                this.trySetHighlightRecord(event.data._id);
             });
     },
 
-    _highlightRecordIfOnList: function (jobId) {
-        if (this.currentView === 'list') {
-            var tr = this.$('tr[g-job-id=' + jobId + ']').addClass('g-highlight');
-            setTimeout(() => tr.removeClass('g-highlight'), 1000);
+    trySetHighlightRecord: function (jobId) {
+        if (this.collection.find((job) => job.id === jobId)) {
```
Comment (brianhelba) [inline plugins/jobs/web_client/views/JobListWidget.js]: [`Collection.get`](http://backbonejs.org/#Collection-get) is the correct (and more efficient) way to look up a Model by id:
```
if (this.collection.get(jobId)) {
```

File: plugins/jobs/web_client/views/JobListWidget.js [hunk 12]
```diff
@@ -217,7 +249,11 @@ var JobListWidget = View.extend({
                 triggerJobClick: this.triggerJobClick,
                 JobStatus: JobStatus,
                 formatDate: formatDate,
-                DATE_SECOND: DATE_SECOND
+                DATE_SECOND: DATE_SECOND,
+                jobCheckedStates: this.jobCheckedStates,
+                jobHighlightStates: this.jobHighlightStates,
+                anyJobChecked: _.find(_.values(this.jobCheckedStates), (status) => status === true),
```
Comment (brianhelba) [inline plugins/jobs/web_client/views/JobListWidget.js]: There's no need to use `_.values` here, Underscore's [`_.find`](http://underscorejs.org/#find) works just fine on objects:
```
anyJobChecked: _.find(this.jobCheckedStates, (status) => status === true),
```

File: plugins/jobs/web_client/views/JobListWidget.js [hunk 13]
```diff
@@ -28,6 +29,29 @@ var JobListWidget = View.extend({
         'change select.g-page-size': function (e) {
             this.collection.pageLimit = parseInt($(e.target).val());
             this.collection.fetch({}, true);
+        },
+        'change input.g-job-checkbox': function (e) {
+            var jobId = $(e.target).closest('tr').attr('g-job-id');
+            if ($(e.target).is(':checked')) {
```
Comment (brianhelba) [inline plugins/jobs/web_client/views/JobListWidget.js]: You usually want to use `e.currentTarget`, to ensure that you're selecting the element this handler was bound to, if the original event fired on a sub-element.

File: plugins/jobs/web_client/views/JobListWidget.js [hunk 14]
```diff
@@ -28,6 +29,29 @@ var JobListWidget = View.extend({
...
+            } else {
+                delete this.jobCheckedStates[jobId];
+            }
+            this._renderData();
+        },
+        'click input.g-job-checkbox-all': function (e) {
+            e.stopPropagation();
+        },
+        'change input.g-job-checkbox-all': function (e) {
+            if ($(e.target).is(':checked')) {
```
Comment (brianhelba) [inline plugins/jobs/web_client/views/JobListWidget.js]: Same here with `e.currentTarget`.

File: plugins/jobs/web_client/views/JobDetailsWidget.js [hunk 15]
```diff
@@ -119,6 +126,19 @@ var JobDetailsWidget = View.extend({
...
+    },
+
+    _cancelJob: function () {
+        const jobId = this.job.id;
+        restRequest({
+            path: `job/${jobId}/cancel`,
+            type: 'PUT',
+            error: null
+        }).done(_.bind(function () {
+            console.log('success');
```
Comment (brianhelba) [inline plugins/jobs/web_client/views/JobDetailsWidget.js]: These debugging statements should be removed, or converted to UI notifications.

File: plugins/jobs/web_client/JobStatus.js [hunk 16]
```diff
@@ -47,12 +53,20 @@ var JobStatus = {
...
+
+            let statusInfo = {
                 text: info.text,
                 icon: info.icon,
                 color: info.color,
+                textColor: 'white',
                 finished: _.has(info, 'finished') ? info.finished : false
             };
+
+            if ('textColor' in info) {
```
Comment (brianhelba) [inline plugins/jobs/web_client/JobStatus.js]: `in` is a bit tricky to use, since it checks up the entire prototype chain. Maybe you want [`_.has`](http://underscorejs.org/#has)?

File: clients/web/src/misc.js [hunk 17]
```diff
@@ -213,7 +213,7 @@ function capitalize(str) {
 
 function splitRoute(route) {
     if (!route) {
-        return {base: '', name: ''};
+        return { base: '', name: '' };
```
Comment (brianhelba) [inline clients/web/src/misc.js]: Not sure why this is changed.

**Comment by cjh1** [comment]
@zachmullen @kotfic Conflicts fixed

File: plugins/jobs/server/models/job.py [hunk 18]
```diff
@@ -384,12 +395,23 @@ def updateJob(self, job, log=None, overwrite=False, status=None,
             job['updated'] = now
             updates['$set']['updated'] = now
 
-            self.update({'_id': job['_id']}, update=updates, multi=False)
+            update_result = self.update(query, update=updates, multi=False)
```
Comment (zachmullen) [inline plugins/jobs/server/models/job.py]: camelCase

File: plugins/jobs/web_client/JobStatus.js (hunk 16 above)
Comment (cjh1) [inline plugins/jobs/web_client/JobStatus.js]: How is it a problem in this situation?

File: plugins/jobs/web_client/views/JobDetailsWidget.js (hunk 15 above)
Comment (cjh1) [inline plugins/jobs/web_client/views/JobDetailsWidget.js]: Can you provide an example of a UI notification? Thanks

File: plugins/jobs/web_client/JobStatus.js (hunk 16 above)
Comment (zachmullen) [inline plugins/jobs/web_client/JobStatus.js]: I think it goes like this:

If you intend to check up the entire prototype chain to see if the object has any property with the given name, use `in`.

If you only want to check if the object itself has the given property (which is likely what you want here) you should use `hasOwnProperty`, which is equivalent to `_.has`.

File: plugins/jobs/web_client/views/JobDetailsWidget.js (hunk 15 above)
Comment (zachmullen) [inline plugins/jobs/web_client/views/JobDetailsWidget.js]: If you're looking for a global notification, it looks like this:

```
events.trigger('g:alert', {
                    icon: 'ok',
                    text: 'Folder info updated.',
                    type: 'success',
                    timeout: 4000
                });
```

File: clients/web/src/misc.js (hunk 17 above)
Comment (cjh1) [inline clients/web/src/misc.js]: Not sure @matthewma7?, seem harmlesss

File: plugins/jobs/server/models/job.py (hunk 18 above)
Comment (cjh1) [inline plugins/jobs/server/models/job.py]: Fixed

File: plugins/jobs/web_client/views/JobListWidget.js (hunk 13 above)
Comment (cjh1) [inline plugins/jobs/web_client/views/JobListWidget.js]: Ok, fixed

File: plugins/jobs/web_client/views/JobListWidget.js (hunk 14 above)
Comment (cjh1) [inline plugins/jobs/web_client/views/JobListWidget.js]: fixed

File: plugins/jobs/web_client/views/JobListWidget.js (hunk 12 above)
Comment (cjh1) [inline plugins/jobs/web_client/views/JobListWidget.js]: Ok, again harmless, but  fixed

File: plugins/jobs/web_client/views/JobListWidget.js (hunk 11 above)
Comment (cjh1) [inline plugins/jobs/web_client/views/JobListWidget.js]: fixed

File: plugins/jobs/web_client/JobStatus.js (hunk 16 above)
Comment (cjh1) [inline plugins/jobs/web_client/JobStatus.js]: Sure, but I am guessing 'textColor' is not in the prototype of {}. I will change it.

File: plugins/jobs/web_client/views/JobDetailsWidget.js (hunk 15 above)
Comment (cjh1) [inline plugins/jobs/web_client/views/JobDetailsWidget.js]: @zachmullen Thanks

Comment (cjh1) [inline plugins/jobs/web_client/views/JobDetailsWidget.js]: Done

File: plugins/jobs/web_client/JobStatus.js (hunk 16 above)
Comment (cjh1) [inline plugins/jobs/web_client/JobStatus.js]: Fixed

Comment (zachmullen) [inline plugins/jobs/web_client/JobStatus.js]: Thanks -- I think this is just something that was hammered into those of us who read Crockford's treatises on javascript: "only use `in` when you really mean it", to the point that when we do see it now, it raises a red flag, even if it probably doesn't matter.

Comment (brianhelba) [inline plugins/jobs/web_client/JobStatus.js]: Yeah, it's not a big deal in this context, but IMO, given that it now raises a mental red flag for me and other Crockford disciples, it decreases the readability of the code every time I see `in` used, because I have to stop and think about whether it's intentional or a bug.

**Comment by cjh1** [comment]
I think I have responded to the review comments

**Comment by zachmullen** [comment]
@kotfic you are the last one with a change request review, can you respond with any additional feedback here?

**Comment by zachmullen** [comment]
I'm going to merge this before it gets out of date again 😎

**Comment by cjh1** [comment]
Awesome, thanks :relieved:

**Comment by mgrauer** [comment]
Nice work, Love Tractor.

**Comment by manthey** [comment]
@cjh1 How do I mark that an ordinary worker job is allowed to be cancelled?  I have a job I wrote previous to this PR that occasionally checks if it was deleted or cancels and terminates early.  The `cancel` function in `plugins/worker/server/__init__.py` prevents this from working, since it cancels the cancel.
